vcd/sample_vcd_test_config.json

```diff
@@ -90,6 +90,44 @@
     "uploadPieceSize": 5,
     "uploadProgress": true
   },
+  "testEnvBuild": {
+    "//": "This section is optional for running tests. It's only needed to build the test environment in a new vCD",
+
+    "// ": "Entities that will be created in an empty vCD use names defined above",
```

Comment:
Also good additional NOTE would be regarding manually clean env if configuration was wrong and script failed.

Reply:
> Also good additional NOTE would be regarding manually clean env if configuration was wrong and script failed.

OK, I just hit this too. I believe this is a _major_ issue and a note is not enough. Maybe I'm wrong, but it looks like the correct way to address this is to keep the Terraform state file, which would magically resolve these issues.

 Example:

1. I made a configuration error and got this error when running `make test-env-build`:
`Error: no provider VDC found with name my-provider-vdc`

2. I fixed the configuration and ran it again, expecting that the environment will *continue* to be built where it left off (in the Terraform fashion), but got this:

```
# terraform apply -auto-approve
vcd_org.myorg: Creating...
vcd_external_network.my-ext-network: Creating...

Error: error creating external network: &errors.errorString{s:"error instantiating a new ExternalNetwork: API Error: 400: [ 743ec78a-a094-45e3-a668-983fa08d3db1 ] External network named 'my-ext-network' already exists."}

  on config.tf line 21, in resource "vcd_external_network" "my-ext-network":
  21: resource "vcd_external_network" "my-ext-network" {

Error: error creating Org: &errors.errorString{s:"error instantiating a new Org: API Error: 400: [ 15413bba-b0e8-4562-aa6d-b09c958c3052 ] The VCD entity myorg already exists."}
```

It looks like state file was lost, because if it were still there Terraform would have reconciled this.

Thanks in advance!

Reply:
p.s. I found that the state file is in `vcd/test-artifacts/buildvcd/` folder. From there I was able to retry building from the point it failed by running `terraform apply`. I suggest the following:
1. Add `make test-env-init` - which would only run `terraform init` and `terraform plan`
2. Leave `make test-env-build` to only `terraform apply`.
3. Rename `test-artifacts/buildvcd/` folder to `test-artifacts/test-env-build/`, so it's easier to find without looking at the source code.

Reply:
> OK, I just hit this too. I believe this is a _major_ issue and a note is not enough. Maybe I'm wrong, but it looks like the correct way to address this is to keep the Terraform state file, which would magically resolve these issues.

This is an unsafe assumption. Terraform can do many things, but not if we provide wrong data, as I have already mentioned previously.

I have hit similar cases when I was preparing the script and tuning it. There is nothing that Terraform guarantees when we fill the template with wrong or missing data. 

Let's be clear on this point. The valid assumption to use this script is that **users have filled the testing part (i.e. everything beyond `testEnvBuild`) with the right data**. Failing to do so would also crash the regular tests spectacularly. We can't have an expectation of better fault tolerance for the build=up task.

Reply:
> * Add `make test-env-init` - which would only run `terraform init` and `terraform plan`
> 
> * Leave `make test-env-build` to only `terraform apply`.
> 
> * Rename `test-artifacts/buildvcd/` folder to `test-artifacts/test-env-build/`, so it's easier to find without looking at the source code.

I don't like making the operation more complicated than what they should be, but I will do that.
Notice, however, that this will not solve hard problems. This script will work just fine when the configuration data has been entered correctly. It will keep failing horribly when users enter invalid data.
Re-applying a configuration script with bad data will only add to the user's blood pressure.